src/main/java/io/fusionauth/http/server/internal/HTTPWorker.java

```diff
@@ -74,114 +82,163 @@ public HTTPWorker(Socket socket, HTTPServerConfiguration configuration, Instrume
+        if (instrumenter != null) {
+          instrumenter.acceptedRequests();
+        }
+
+        httpInputStream = new HTTPInputStream(configuration, request, inputStream, bodyBytes);
+        request.setInputStream(httpInputStream);
+
+        // Set the Connection response header as soon as possible
+        // - This needs to occur after we have parsed the pre-amble so we can read the request headers
+        response.setHeader(Headers.Connection, request.isKeepAlive() ? Connections.KeepAlive : Connections.Close);
+
+        // Ensure the preamble is valid
+        Integer status = validatePreamble(request);
+        if (status != null) {
+          closeSocketOnError(response, status);
+          return;
+        }
+
+        // Handle the Expect: 100-continue request header.
         String expect = request.getHeader(Headers.Expect);
         if (expect != null && expect.equalsIgnoreCase(HTTPValues.Status.ContinueRequest)) {
           state = State.Write;
 
-          // If the "expect" wasn't accepted, close the socket and exit
-          if (!expectContinue(request)) {
-            close(Result.Success, response);
+          boolean doContinue = handleExpectContinue(request);
+          if (!doContinue) {
+            // Note that the expectContinue code already wrote to the OutputStream, all we need to do is close the socket.
+            closeSocketOnly(CloseSocketReason.Expected);
             return;
           }
 
           // Otherwise, transition the state to Read
           state = State.Read;
         }
 
-        if (request.isKeepAlive()) {
-          response.setHeader(Headers.Connection, Connections.KeepAlive);
-          keepAlive = true;
-        } else {
-          response.setHeader(Headers.Connection, Connections.Close);
-          keepAlive = false;
+        // Transition to processing
+        state = State.Process;
+        logger.trace("[{}] Set state [{}]. Call the request handler.", Thread.currentThread().threadId(), state);
+        configuration.getHandler().handle(request, response);
+        logger.trace("[{}] Handler completed successfully", Thread.currentThread().threadId());
+
+        // Drain the InputStream before closing the HTTP Response
+        long startDrain = System.currentTimeMillis();
+        int drained = httpInputStream.drain();
+        if (drained > 0 && logger.isTraceEnabled()) {
+          long drainDuration = System.currentTimeMillis() - startDrain;
+          logger.trace("[{}] Drained [{}] bytes from the InputStream. Duration [{}] ms.", Thread.currentThread().threadId(), drained, drainDuration);
         }
 
-        logger.debug("Calling the handler.");
-        var handler = configuration.getHandler();
-        state = State.Process; // Transition to processing
-        handler.handle(request, response);
+        // TODO : Daniel : Review : If we must drain the InputStream prior to calling close, it seems like this would be a problem since the request handler
```

Comment:
Would like to discuss this with some other folks. It seems to me that we must drain the `InputStream` prior to calling `close()` on the `OutputStream`. 

I can demonstrate the effect of moving the `drain` past the `close()`. 

In practice, I think this is ok - however - because we can't control what the request handler does with `close()` this seems brittle. 

We could remove the ability for a request handler to call `close()` as one option so we control it.